craft_application/git/_git_repo.py

```diff
@@ -308,3 +405,46 @@ def _resolve_ref(self, ref: str) -> str:
             ) from error
         logger.debug("Resolved reference %r for name %r", reference, ref)
         return reference
+
+    @classmethod
+    def clone_repository(
+        cls,
+        *,
+        url: str,
+        path: Path,
+        checkout_branch: str | None = None,
+        depth: int = 0,
+    ) -> GitRepo:
+        """Clone repository to specific path and return GitRepo object.
+
+        :param url: the URL to clone repository from
+        :param path: path to the directory where repository should be cloned
+        :param checkout_branch: optional branch which should be checked out
+
+        :returns: GitRepo wrapper around the repository
+
+        raises GitError: if cloning repository cannot be done
+        """
+        if is_repo(path):
+            raise GitError("Cannot clone to existing repository")
+
+        logger.debug("Cloning %s to %s", url, path)
+        if checkout_branch is not None:
+            logger.debug("Checking out to branch: %s", checkout_branch)
+
+        try:
+            # TODO: consider using git CLI with --single-branch
+            #  as it is not implemented in pygit2
+            #  https://github.com/libgit2/pygit2/issues/1290
```

Comment:
I'm not sure this is worth switching to the CLI version of git for, but I like having the reference to the issue so we can implement that once pygit2 implements it.

Could you make a feature request for `--single-branch`, tag it as blocked, and reference the upstream issue?

Reply:
Resolving due to https://github.com/canonical/craft-application/pull/396#discussion_r1690252171